content/asa/0002-teorema-mestre.md

```diff
@@ -0,0 +1,522 @@
+
+Ficamos então com o somatório:
+
+$$
+\sum_{k=0}^{\log_b n} O \left( a^k (\frac{n}{b^k})^d \right) =
+O \left( n^d \sum_{k=0}^{\log_b n} \left( \frac{a}{b^d} \right)^k \right)
+$$
+
+Analisemos agora caso a caso:
+
+- No caso 1, temos que $d < \log_b a \Leftrightarrow b^d < a \Leftrightarrow \frac{a}{b^d} > 1$.
+  Neste caso, o somatório explode e é dominado pelo último termo, ou seja a complexidade tem _upper bound_ de
+  $$
+  O \left( n^d \left( \frac{a}{b^d} \right)^{\log_b n} \right) =
+  O \left( n^d \frac{a^{\log_b n}}{n^d} \right) =
+  O \left( a^{\log_b n} \right)
+  $$
+- No caso 2, ficamos com
+  $$ O \left( n^d (log_b n + 1) \right) = O \left( n^d \log n \right) $$
+- No caso 3, o somatório é majorado por uma série que converge (uma vez que $d > \log_b a \Leftrightarrow b^d > a \Leftrightarrow \frac{a}{b^d} < 1$) pelo que
+  $$ O \left( n^d \right) $$
+
+:::
+
+:::details[Exemplo 1]
+
+Tenhamos $T(n) = T(\frac{n}{3}) + n$.
+Aqui, podemos dizer que:
+
+$$
+a = 9, \quad b = 3, \quad d = 1
+$$
+
+Daqui, obtemos que
+
+$$
+\log_{b}{a} = \log_{3}{9} = 2 > 1 = d
+$$
+
+Assim sendo, admitimos que estamos na presença do [**caso 1**](color:orange), e que, portanto,
+
+$$
+T(n) = O(n^{\log_{b}{a}}) = O(n^2).
+$$
+
+:::
+
+:::details[Exemplo 2]
+
+Tenhamos:
+
+```cpp
+int f(int n) {
+  int i, j;
+  i = 0;
+  j = 0;
+  while (i < n) {
+    j++;
+    i += 2;
+  }
+  if (n > 1) {
+    i = 2*f(j) + f(j);
+  }
+  return i;
+}
+```
+
+Podemos procurar ver como o ciclo se comporta (e qual é a sua condição de paragem) recorrendo ao [**método da tabela**](color:green). Neste caso:
+
+| $k$ | $i$  | $j$ |
+| --- | ---- | --- |
+| $0$ | $0$  | $0$ |
+| $1$ | $2$  | $1$ |
+| $2$ | $4$  | $2$ |
+| ... | ...  | ... |
+| $k$ | $2k$ | $k$ |
+
+Onde $k$ é a variável de controlo do loop (conta o número de iterações), e $i$ e $j$ são variáveis que vão sendo atualizadas durante o mesmo. Podemos notar que $k = 0$ corresponde ao momento exatamente antes do loop.
+
+Ora, podemos perceber que $i$ vai crescendo com aspeto $2k$ - isto é, vai sempre sendo igual ao dobro do número de iterações atual do ciclo. Além disso, através da tabela, podemos também observar que, no fim do ciclo, $j = k$.
+
+Temos que o ciclo para com $i \geq n$ - ou seja, quando
+
+$$
+2k \geq n \leftrightarrow k \geq \frac{n}{2}
+$$
+
+Podemos, então, dizer que o ciclo corre $\frac{n}{2}$ vezes, e que a sua complexidade temporal é $O(n)$.
+
+Como observado, $j = k$ no fim do ciclo, pelo que a chamada recursiva da função (dentro do bloco `if`) é realizada com [$j = \frac{n}{2}$](color:orange) [**duas vezes**](color:yellow) (a função recursiva é chamada duas vezes), pelo que podemos admitir que essa secção leva $2\cdot T(\frac{n}{2})$. A função $T(n)$ total da função corresponde, portanto, a
+
+$$
+T(n) = \textcolor{yellow}{2}\cdot T\left(\textcolor{orange}{\frac{n}{2}}\right) + O(n).
```

Comment:
Colors on KaTeX have a problem that they're not dynamic, that is, they don't adapt to the light/dark theme of the website. I have a solution in mind for this, I'll fix it myself.